### PR TITLE
[4.0] Modulefilterfix

### PR DESCRIPTION
**Pull request by astridx**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/30302 .

### Summary of Changes

This pull request turns the links `site modules` and `admin modules` into one link `modules`. So the URI stays the same when changing from site module to admin module. This has the advantage that the filters, especially reset filters, continue to work correctly.

> Note If the distinction  between admin and site modules is wanted, it would be better to include the `client_id `in a hidden form field. But I would do this in a separate PR.

### Testing Instructions

1. Apply this patch. You can use patch tester. `npm` or `composer` are **not** required.

2. Open the modules and check that admin and site modules view can be changed with the drop down and that the modules entry is always active in the admin menu on the left side.

Modules Site test Administration

3. Set filters, switch between admin and custom and check that the filters are deleted when you switch. See https://github.com/joomla/joomla-cms/issues/30302
4. Check, that quicktask near the content menu item works correct.

Content Dashboard test Administration

5. Use different presets and check, that these work correct.

Modules Administrator Menu test Administration

### Actual result BEFORE applying this Pull Request

Reset Filter does not work properly, see https://github.com/joomla/joomla-cms/issues/30302.

### Expected result AFTER applying this Pull Request

Reset Filter does work.

### Documentation Changes Required

No

### CODE REVIEW

**Comment by paternax** [comment]
I have tested this item :white_check_mark: successfully on e1a1de0db58027a02475409c68c9099c2bc3123b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30577">issues.joomla.org/tracker/joomla-cms/30577</a>.</sub>

**Comment by brianteeman** [comment]
This PR results in the loss of the quicktask link for creating a new module. I would be sad to see this go

**Comment by Quy** [comment]
Quick icon needs to be updated.

File: administrator/language/en-GB/mod_menu.ini [hunk 1]
```diff
@@ -33,8 +33,6 @@ MOD_MENU_EXTENSIONS_EXTENSION_MANAGER="Manage"
 MOD_MENU_EXTENSIONS_EXTENSIONS="Extensions"
 MOD_MENU_EXTENSIONS_LANGUAGE_MANAGER="Languages"
 MOD_MENU_EXTENSIONS_MODULE_MANAGER="Modules"
-MOD_MENU_EXTENSIONS_MODULE_MANAGER_ADMINISTRATOR="Administrator Modules"
```
Comment (infograf768) [inline administrator/language/en-GB/mod_menu.ini:36]: We do need this for the alternate.xml
If not we get
Screen Shot 2020-09-06 at 09 41 47

EDIT: see my conclusion below

Comment (infograf768) [reply administrator/language/en-GB/mod_menu.ini:36]: Also, I re-added the quicktask in alternate.xml to create new site modules
code
```
		<menuitem
			title="MOD_MENU_EXTENSIONS_MODULE_MANAGER"
			quicktask-title="COM_CONTENT_MENUS_NEW_SITE_MODULE"
			type="component"
			element="com_modules"
			link="index.php?option=com_modules&amp;view=modules"
			quicktask="index.php?option=com_modules&amp;view=select&amp;client_id=0"
		/>
```
to get
Screen Shot 2020-09-06 at 09 49 25

The problem is that if we keep the administrator modules menu item, we still have the issue when switching from site to administrator via the menu items.

Also if we keep some quicktask for any of these modules s menu items, then the last quicktask used, if we create a new module or just look at the page and click on one of the modules menu items again, defines which client modules will be displayed next, therefore same issue with filtering.

Comment (infograf768) [reply administrator/language/en-GB/mod_menu.ini:36]: Conclusion: if we keep these changes, the administrator modules menu item has also to be deleted on alternate.xml and all related quicktask killed.

**Comment by infograf768** [comment]
I took off the administrator modules menu item from alternate.xml
Then, I used the Modules menu item and chose Administrator as client, then New:

Although the correct Modules list is displayed, the manager title is always `Modules (Site)` instead of `Modules (Administrator)`

**Comment by astridx** [comment]
> @brianteeman This PR results in the loss of the quicktask link for creating a new module. I would be sad to see this go

Yes you are right. This is a pity.
But, if there is only one link in the side menu, quicktask would be confusing, right? You then don't know whether a `site` or `admin` module is being created. Do you have a suggestion? Of course it is easier with two `menu item entries` (one for site and one for admin). But the way I see it, this is only possible with a GET variable. With a GET variable in the URL, however, any filters that may have been set are not deleted. Or do you see a possibility there?

**Comment by astridx** [comment]
> @Quy  Quick icon needs to be updated.
> 
> 


Thanks, I missed that. Now I've corrected it.

**Comment by astridx** [comment]
> @SharkyKZ On glip your wrote, that there is no need to remove site/client module URLs, because this can be fixed in the model alone.

Thanks for checking. I have tried this https://github.com/joomla/joomla-cms/pull/30323 and explained [here](https://github.com/joomla/joomla-cms/issues/30302#issuecomment-671816892) why I think this is in my opinion not possible. What am I missing? Can you give me a hint how I can implement it?

**Comment by astridx** [comment]
> @infograf768 I took off the administrator modules menu item from alternate.xml
> Then, I used the Modules menu item and chose Administrator as client, then New:
> 
> Although the correct Modules list is displayed, the manager title is always `Modules (Site)` instead of `Modules (Administrator)`

Thanks, I missed `alternate.xml`. Now I've corrected it.

Can you explain to me in more detail which `title` is incorrect? It looks right for me. See pictures. But maybe I get you wrong.

Modules Site test Administration(1)
Modules Administrator test Administration

**Comment by Quy** [comment]
> But, if there is only one link in the side menu, quicktask would be confusing, right?

Make it same as the Modules quick icon.

**Comment by infograf768** [comment]
@astridx 
> Can you explain to me in more detail which title is incorrect? It looks right for me. See pictures. But maybe I get you wrong.

modules_filter

**Comment by SharkyKZ** [comment]
@astridx I don't have an exact solution yet. Using client ID from request when building context key or clearing filters when client ID has changed could work. In any case, we have more instances like this (e.g. template styles).

**Comment by infograf768** [comment]
> clearing filters when client ID has changed could work

That would be ideal so we can keep quicktasks.

**Comment by paternax** [comment]
I have tested this item :white_check_mark: successfully on abb778f1a206a3a966732d1df59c070312a46d55<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30577">issues.joomla.org/tracker/joomla-cms/30577</a>.</sub>

**Comment by ceford** [comment]
I am not sure about this. When I change from Site to Administrator I see the filter being cleared before page load. That is probably as annoying as not clearing the filters. Otherwise, it seems to work.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30577">issues.joomla.org/tracker/joomla-cms/30577</a>.</sub>

**Comment by astridx** [comment]
> > But, if there is only one link in the side menu, quicktask would be confusing, right?
> 
> @Quy Make it same as the Modules quick icon.


I inserted quicktast again in all presets. like you suggested.

**Comment by astridx** [comment]
> @ceford I am not sure about this. When I change from Site to Administrator I see the filter being cleared before page load. That is probably as annoying as not clearing the filters. Otherwise, it seems to work.
> This comment was created with the [J!Tracker Application](https://github.com/joomla/jissues) at [issues.joomla.org/tracker/joomla-cms/30577](https://issues.joomla.org/tracker/joomla-cms/30577).

If I understand you correctly, then it has to be like this, otherwise wrong data would be loaded (or they have to be loaded twice). The elements in the list are loaded using the filters. Did I understand you correctly?

**Comment by astridx** [comment]
@infograf768 
> Although the correct Modules list is displayed, the manager title is always `Modules (Site)` instead of `Modules (Administrator)`

Thank you. I corrected the title.

**Comment by astridx** [comment]
> @SharkyKZ  astridx I don't have an exact solution yet. Using client ID from request when building context key or clearing filters when client ID has changed could work. In any case, we have more instances like this (e.g. template styles).

But isn't it also wrong with  template styles? See
https://github.com/joomla/joomla-cms/issues/30302#issuecomment-672791829

**Comment by Quy** [comment]
In the quick icon, the tooltip for the + icon is "Add Site Module" which is not correct depending on the current module type selected.

**Comment by astridx** [comment]
> @Quy uy In the quick icon, the tooltip for the + icon is "Add Site Module" which is not correct depending on the current module type selected.

Thanks done.

**Comment by SharkyKZ** [comment]
> But isn't it also wrong with template styles? See
> [#30302 (comment)](https://github.com/joomla/joomla-cms/issues/30302#issuecomment-672791829)

Yes, it's wrong. But this should be fixed instead of being stripped out.

**Comment by ghost** [comment]
> 5. Use different presets and check, that these work correct.

I have tested this item :red_circle: unsuccessfully on 00f8d521a28338af353b286c7c6424e34a819889<br /><br />
1. Go to ```Menu Dashboard```
1. Click on ```Add module to the dashboard```
1. Select a Module Type
1. At Top of Modal-Windows Part of Admin-Template and at Bottom Buttons ```Close```, ```Save & Close``` are shown. Click on ```Save & Close``` at Bottom didn't work:

Screenshot_2020-09-11 Menu Dashboard - fw4 - Administration



#### System Information
php: Linux lamp131.cloudaccess.net 3.10.0-714.10.2.lve1.5.17.el6h.x86_64 #1 SMP Mon Apr 30 13:48:47 EDT 2018 x86_64
dbserver: mysql
dbversion: 5.7.24-cll-lve
dbcollation: utf8_general_ci
dbconnectioncollation: utf8mb4_general_ci
dbconnectionencryption: 
dbconnencryptsupported: false
phpversion: 7.3.21
server: Apache
sapi_name: cgi-fcgi
version: Nightly Build - Joomla! 4.0.0-beta4-dev Development [ Mañana ] 29-July-2020 18:21 GMT
useragent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:80.0) Gecko/20100101 Firefox/80.0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30577">issues.joomla.org/tracker/joomla-cms/30577</a>.</sub>

**Comment by paternax** [comment]
> 
> I have tested this item 🔴 unsuccessfully on [00f8d52](https://github.com/joomla/joomla-cms/commit/00f8d521a28338af353b286c7c6424e34a819889)
> 
>     1. Go to `Menu Dashboard`
> 
>     2. Click on `Add module to the dashboard`
> 
>     3. Select a Module Type
> 
>     4. At Top of Modal-Windows Part of Admin-Template and at Bottom Buttons `Close`, `Save & Close` are shown. Click on `Save & Close` at Bottom didn't work:
> 

I can't confirm this problem. The Save&Close worked in the scenario Formation h. described.

**Comment by paternax** [comment]
I have tested this item :white_check_mark: successfully on 00f8d521a28338af353b286c7c6424e34a819889<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30577">issues.joomla.org/tracker/joomla-cms/30577</a>.</sub>

**Comment by infograf768** [comment]
I confirm @Formatio-hippocampi findings
This PR breaks inserting a module when using the bottom buttons.
It works when using the toolbar top buttons but does not display at once. Page has to be reloaded to see the new module.

Displaying the status and the toolbar in the modal window is wrong.

Screen Shot 2020-09-11 at 10 05 02

I have tested this item :red_circle: unsuccessfully on 00f8d521a28338af353b286c7c6424e34a819889<br />

**Comment by infograf768** [comment]
It looks like the  issue comes from
/administrator/components/com_modules/tmpl/select/default.php line 61

I solved it by using
`<?php $link = 'index.php?option=com_modules&task=module.add&eid=' . $item->extension_id . $this->modalLink; ?>`

instead of
`<?php $link = 'index.php?option=com_modules&task=module.add&eid=' . $item->extension_id; ?>`

**Comment by astridx** [comment]
Thank you  @Formatio-hippocampi and @infograf768  . I did not see that. I have corrected it.
A module can now also be correctly added via "add module to the dashboard".

Control Panel test Administration

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 17da55f99d6ab5c3791eaeed725a8d431c35df09<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30577">issues.joomla.org/tracker/joomla-cms/30577</a>.</sub>

**Comment by infograf768** [comment]
Works OK now but I still think that clearing filters when client ID has changed would be a better solution.

**Comment by pabloarias** [comment]
I have tested this item :white_check_mark: successfully on 17da55f99d6ab5c3791eaeed725a8d431c35df09<br /><br />Hello:

I've followed Testing Instructions, first without applying the path and then applying it. Now, when I back to modules list, I see modules.

Tested with Joomla Beta 5-dev and PHP 7.3.22.

Thank you!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30577">issues.joomla.org/tracker/joomla-cms/30577</a>.</sub>

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 17da55f99d6ab5c3791eaeed725a8d431c35df09<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30577">issues.joomla.org/tracker/joomla-cms/30577</a>.</sub>

**Comment by brianteeman** [comment]
## Before
before


## After
after

**Comment by astridx** [comment]
@brianteeman
Thank you for testing. Can you explain in short words what you mean?

**Comment by brianteeman** [comment]
as shown in the video if the last module manager page was with the admin modules selected then the add new module on the dashboard will offer to create a new admin module instead of a site module

**Comment by astridx** [comment]
@brianteeman
> as shown in the video if the last module manager page was with the admin modules selected then the add new module on the dashboard will offer to create a new admin module instead of a site module


I like this. Why do you think it is a failure?

**Comment by brianteeman** [comment]
Simple. A link must always go to the same predictable place. If nothing else it is an accessibility failure and an unexpected behaviour that a link doesnt go where you think it will go because of an action you performed 4 hours earlier.

**Comment by astridx** [comment]
OK.  I am no a11y expert. Then I close this here

**Comment by pabloarias** [comment]
Hello:

I understand Brian, a link should be predictable, OK. But current behaviour, an empty list after using some filters (issue #30302), is not correct.

I think this solution by @astridx is the best we have right now. User has played with the filters by himself, so the link has not an terrible accessibility issue, user goes to a predictable place for him. I think it's the behaviour J3 has since years ago and it's useful.

Thank you and best regards

**Pull request closed without merging**